### ISSUE TITLE
Make as many services private as possible

### ISSUE THREAD

**Issue by nicolas-grekas** [symfony/symfony#23822]
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| BC Break report? | no
| RFC?             | no
| Symfony version  | 3.4

There are many services which are public right now but have not reason to be.

**Comment by javiereguiluz** [comment]
Can we make a list of those services so the community can submit PRs making them private ... or do you prefer to work on this yourself? Thanks.

**Comment by Simperfit** [comment]
I think we can make a list, it will a lot of work for only one person :)

**Comment by javiereguiluz** [comment]
In the Symfony Demo upgraded to Symfony 3.3.6 and Symfony Flex I see a lot of services that could be removed. For example, most of these services related to translation:

| Service ID               | Service class
| ------------------------ | --------------------------------------------------------
| translation.dumper.csv   | Symfony\Component\Translation\Dumper\CsvFileDumper
| translation.dumper.ini   | Symfony\Component\Translation\Dumper\IniFileDumper
| translation.dumper.json  | Symfony\Component\Translation\Dumper\JsonFileDumper
| translation.dumper.mo    | Symfony\Component\Translation\Dumper\MoFileDumper
| translation.dumper.php   | Symfony\Component\Translation\Dumper\PhpFileDumper
| translation.dumper.po    | Symfony\Component\Translation\Dumper\PoFileDumper
| translation.dumper.qt    | Symfony\Component\Translation\Dumper\QtFileDumper
| translation.dumper.res   | Symfony\Component\Translation\Dumper\IcuResFileDumper
| translation.dumper.xliff | Symfony\Component\Translation\Dumper\XliffFileDumper
| translation.dumper.yml   | Symfony\Component\Translation\Dumper\YamlFileDumper
| translation.extractor    | Symfony\Component\Translation\Extractor\ChainExtractor
| translation.extractor.php| Symfony\Bundle\FrameworkBundle\Translation\PhpExtractor
| translation.loader       | Symfony\Bundle\FrameworkBundle\Translation\TranslationLoader
| translation.loader.csv   | Symfony\Component\Translation\Loader\CsvFileLoader
| translation.loader.dat   | Symfony\Component\Translation\Loader\IcuDatFileLoader
| translation.loader.ini   | Symfony\Component\Translation\Loader\IniFileLoader
| translation.loader.json  | Symfony\Component\Translation\Loader\JsonFileLoader
| translation.loader.mo    | Symfony\Component\Translation\Loader\MoFileLoader
| translation.loader.php   | Symfony\Component\Translation\Loader\PhpFileLoader
| translation.loader.po    | Symfony\Component\Translation\Loader\PoFileLoader
| translation.loader.qt    | Symfony\Component\Translation\Loader\QtFileLoader
| translation.loader.res   | Symfony\Component\Translation\Loader\IcuResFileLoader
| translation.loader.xliff | Symfony\Component\Translation\Loader\XliffFileLoader
| translation.loader.yml   | Symfony\Component\Translation\Loader\YamlFileLoader
| translation.writer       | Symfony\Component\Translation\Writer\TranslationWriter
| translator               | alias for "translator.data_collector"
| translator.default       | Symfony\Bundle\FrameworkBundle\Translation\Translator
| translator_listener      | Symfony\Component\HttpKernel\EventListener\TranslatorListener

**Comment by stof** [comment]
there is an issue about doing this in 3.4: if they get referenced only once, they will be inlined, which forbids accessing them through `get` even through the BC layer. And this is a BC break.

Btw, we don't have any upgrade path for switching a service to private (be it a core service of an internal services). I suggest adding a feature for that.
Public services and aliases could get a new flag on them (no idea for the name though), which would signal Symfony that we plan to switch it to private in the future. Then, when doing a `get` on this service, the container would first check if the id is in the list of such services and trigger a deprecation warning.
Doing it in `get` would mean that it adds some overhead to all `get` calls. I don't know how to improve this (especially as getting a -to-become-private public service may not trigger the instantiation if already done)

**Comment by stof** [comment]
@javiereguiluz The `translator` alias should stay public IMO, as it is the main API of the translation system.
About `translation.loader `translation.writer` and `translation.extractor`, I'm not sure. All others can indeed become private (they were public for pre-3.3 lazy-loading)

**Comment by nicolas-grekas** [comment]
@stof there is another option, which is much lighter: we create a special public service but deprecated, that references them all.
Then we can safely make them private so that people will get a notice in 3.4.

**Comment by stof** [comment]
@nicolas-grekas note that the feature I'm talking about to deprecate public access to services should not go away in 4.0, as 4.x+ of Symfony may still want to do such changes in the future (and then, we don't even have the private get BC layer for private services referenced multiple times), and external bundles will have the same need too (once they start dropping support for Symfony < 3.4, most of their services won't need to be public either)

Your proposal about using a special deprecated public services is not fine:

- it actually requires 2 such services (for cases of public services referenced nowhere that you want to make private)
- it only works for Symfony 3.3 and 3.4. Once people are on Symfony 4, it does not work. so it is not a solution for bundles (bundles won't drop support for Symfony < 3.4 until after the Symfony 4 release, so they need a migration path compatible with Symfony 4)

**Comment by nicolas-grekas** [comment]
Of course, I was seeking for the laziest plan :)
If anyone wants to implement the feature, PR welcomed ;)

**Comment by sstok** [comment]
To really make translations loading lazy we'd need something like this https://github.com/symfony/symfony/pull/17563 right?

**Comment by chalasr** [comment]
👍  for introducing what @stof suggests, it would be useful for many third party bundles for sure.

**Comment by nicolas-grekas** [comment]
Here is the plan:
- add a tag (`container.private`?) that express the private-in-next-major state
- bundles would keep these services as public for <=3.3 compat, but we would turn them private because we can
- add a compiler pass that turns these into private, with double registration trick to trigger the existing deprecation about private services no being fetchable anymore in 4.0
- in 4.0, this compiler pass would just turn these services to private
- in 4.1, we deprecate the `container.private` tag and the corresponding pass

That would allow bundles to still have the services public with <3.3, trigger the deprecation with 3.4, and have the services really private when used with 4.0.

(I'm still trying to *not* have to create and maintain a feature to move from public to private on the long term)

**Comment by nicolas-grekas** [comment]
Some comments posted on #23867 (which is now closed):

From @Tobion:
> Just some thoughts: The whole concept of private services is only to shrink the container. So basically it's just implementation detail and ideally as developer I should not need to care about this.
So IMO in an ideal world there is no such thing as private and public services.
> 
> Can't we achieve that goal? Now that the container creates files per factory, the container should scale better with the amout of services, right? Can't we improve that even more so that the overhead per service is negligible? E.g. less internal tracking of serivce ID mapping to file name/method name.
Then we don't care about private or public at all.
> 
> I think the only difference are inlined services. If we'd say everything is public, there is no private anymore, the container would need to keep a reference to the inlined services. But does this really matter?

Then from @fabpot:
> I like @Tobion thoughts a lot :)

Then me:
> In Flex, we made the move of loading everything in the src/ folder because services where private by default - so they would be cleaned up anyway.

**Comment by stof** [comment]
The difference for private services is not only the inlining, but also the removal indeed (as the container is able to ensure that they won't ever be needed, as they are forbidden for dynamic retrieval).
So I don't think we can remove this distinction

**Comment by Tobion** [comment]
@stof if there was no performance difference between removing or not removing private services, then the distinction is not neccessary.

**Comment by sroze** [comment]
My thoughts after trying `dev-master` on the symfony-demo project... I think this is asking for troubles. On such a small project, I had do these 2 things to have the application working:
- Update SymfonyBundle's demo to have the `doctrine` service accessible from the `ControllerTrait`
- Update pagerfanta's `white_october_pagerfanta.view_factory` service to `public="true"`

This means almost all the Symfony libraries will have to be updated for Symfony 4... Did we measure what is the performance improvement on such change that would justify this? :)

**Comment by nicolas-grekas** [comment]
It's not only about perf, it's also a lot about design. Bundles don't care about this public/private difference, so that everything is usually public, thus open to service location.
All bundles *should* have made most their services private from the beginning (as we try to do in core). But having everything public by default just makes this a dream.
Turning to private by default fixes this.
Time now to submit PRs to bundles and make them either explicitly opt-in for public on a case by case basis, or have them leverage (new DI ways of providing) regular injection.

**Comment by stof** [comment]
@sroze the fact that the default visibility of services is changing indeed means that bundles have to be updated. But they can be updated in 2 different ways (depending on the services):

- marking them as public explicitly if it is actually part of the 
- use new Symfony DI 3.3+ feature to perform lazy-loading without requiring services to be public.

**Comment by sroze** [comment]
@fabpot I guess you wanted to tag @stof on the last sentence.

**Comment by stof** [comment]
Regarding the error message being difficult to understand, this is something we could work on during the stabilisation phase.

**Comment by sroze** [comment]
I agree with the idea that it would be nice to have libraries providing clear extension points (I guess that's what you are trying to solve with private services by default).

Still, if the main reasons are nice-to-have design-thinking and not-proven performances improvements, I guess that's still a debatable choice to force all the existing ecosystem to update. I'd go as far as saying that now that we have timely-based releases, it's not because we can break the BC at each major version that we should 😛

I'd follow on @fabpot's argument: private services were never a communicated best practice. I'd much more prefer a documentation approach (and also "blaming" libraries exposing too much) than enforcing it at the framework level, or we are going to have unhappy users all the way from 3.x to 4.x.

**Comment by nicolas-grekas** [comment]
Don't take me wrong, design is not a nice-to-have: it's something ppl expect their framework to help them do better.
About the performance argument, having services private allows 2 things: inlining, and removals. Both contribute to generate a very clean generated container, and that's also what I'd expect from a state of the art compiled container (let's say we pretend being one :))
But if nobody cares, let's drop the private/public distinction. The performance impact won't be high, and if ppl do `$container->get()` all over, that's their mistake. We can do that, I'd have no issue with that if that's what we decide. But we cannot stay in between, we have to move forward in one direction or another, and be core to our values.

> it's not because we can break the BC at each major version that we should

I'm telling this all the day while reviewing PRs so I know that for sure. Here making services public in an explicit way is VERY easy, so the required fix is trivial. That's not an excuse, but a balanced argument: if it's easy to do while improving the architecture of all SF apps, that's worth it.
At least that's the motivation behind the current state.
